R/variable.R

```diff
@@ -211,8 +211,17 @@ FactorToNumeric <- function(x, binary = TRUE, name = NULL, remove.first = TRUE)
 FactorToIndicators <- function(variable, name = NULL)
 {
     if (is.null(name))
-        name = Names(variable)
-    result <- stats::model.matrix( ~ variable - 1)
+        name <- Names(variable)
+    if (length(levels(variable)) == 1) { 
+        # When only a single level, model.matrix will fail.
+        # Instead, create a single-column matrix where all
+        # non-na values are set to 1. 
+        result <- matrix(as.numeric(!is.na(variable)), ncol = 1)
+        result[is.na(variable)] <- NA
```

Comment:
calling `is.na` twice here. Could be more efficient with the following?
```suggestion
        result <- as.matrix(as.integer(variable))
```

Reply:
I suppose another additional danger here is when there are unobserved levels but unsure how this gets handled when this function is used anyway. E.g. `variable <- factor(rep(c("B", NA), 5:6), levels = LETTERS[1:3])`

Reply:
It seems to handle unobserved levels just fine down the normal code path.

as.integer doesn't work here because I need the non-na value to be 1 and as.integer can generate other values if the sole observed level is not the first one.

Reply:
It'll never enter this code block unless it has only 1 level though.

Reply:
i.e. if there are two levels (with one observed and one not observed) then it'll `nlevels(x)`  is 2

Reply:
There is no way of knowing which ones are observed or not without a call to `table` or `unique`.

Reply:
Yes the old code only hit a problem when nlevels(x) == 1. If there are two levels but one is not observed then the old code is fine.

Reply:
oh i see what you mean now